hawkular-command-gateway/hawkular-command-gateway-api/src/test/java/org/hawkular/cmdgw/api/ApiDeserializerTest.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    @Test
+    public void testGenericSuccessResponse() {
+        GenericSuccessResponse newpojo;
+        GenericSuccessResponse pojo = new GenericSuccessResponse();
+        pojo.setMessage("howdy!");
+
+        newpojo = testSpecificPojo(pojo);
+        Assert.assertEquals(pojo.getMessage(), newpojo.getMessage());
+    }
+
+    @Test
+    public void testGenericErrorResponse() {
+        GenericErrorResponse newpojo;
+        GenericErrorResponse pojo = new GenericErrorResponse();
+        pojo.setErrorMessage("howdy!");
+        pojo.setStackTrace("stack trace here");
+
+        newpojo = testSpecificPojo(pojo);
+        Assert.assertEquals(pojo.getErrorMessage(), newpojo.getErrorMessage());
+        Assert.assertEquals(pojo.getStackTrace(), newpojo.getStackTrace());
+    }
+
+    @Test
+    public void testEchoRequest() {
+        EchoRequest newpojo;
+        EchoRequest pojo = new EchoRequest();
+        pojo.setEchoMessage("howdy!");
+
+        newpojo = testSpecificPojo(pojo);
+        Assert.assertEquals(pojo.getEchoMessage(), newpojo.getEchoMessage());
+    }
+
+    @Test
+    public void testEchoResponse() {
+        EchoResponse newpojo;
+        EchoResponse pojo = new EchoResponse();
+        pojo.setReply("what up?");
+
+        newpojo = testSpecificPojo(pojo);
+        Assert.assertEquals(pojo.getReply(), newpojo.getReply());
+    }
+
+    @Test
+    public void testMessageWithExtraData() {
+        // serialize a message and some extra data
+        final String testMessage = "this is the message";
+        final String testExtraData = "this is extra data";
+
+        GenericSuccessResponse msg = new GenericSuccessResponse();
+        msg.setMessage(testMessage);
+        ByteArrayInputStream extraData = new ByteArrayInputStream(testExtraData.getBytes());
+        BinaryData fullData = ApiDeserializer.toHawkularFormat(msg, extraData);
+
+        // now deserialize the data
+        ApiDeserializer ad = new ApiDeserializer();
+        BasicMessageWithExtraData<GenericSuccessResponse> deserializedFullData = ad.deserialize(fullData);
+        GenericSuccessResponse deserializedMessage = deserializedFullData.getBasicMessage();
+        BinaryData deserializedExtraData = deserializedFullData.getBinaryData();
+        String deserializedExtraDataString = new Scanner(deserializedExtraData, "UTF-8").useDelimiter("\\A").next();
+
+        // make sure the deserialized data matches what we serialized
+        Assert.assertEquals(testMessage, deserializedMessage.getMessage());
+        Assert.assertEquals(testExtraData, deserializedExtraDataString);
+
+    }
+
+    @Test
+    public void testMessageWithLargeExtraData() throws Exception {
+        // serialize a message and some extra data
+        final String testMessage = "this is the message";
+        final String testExtraData;
+
+        // stream [0, 1, 2, ..., 255], repeat that 10 times
+        byte[] bytes = new byte[0xff * 10];
```

Comment:
By the way, isn't there an off-by-one error ? `[0,...,255]` has a length of 256, no?